tests/integration/test_task.py

```diff
@@ -80,3 +82,219 @@ def _run(self, input_files):
     print(order)
     assert len(order) == 12
     assert order == sorted(order)
+
+
+@pytest.mark.parametrize('executables', ['ls', 'mv'])
+def test_which(executables):
+    """Test the which wrapper."""
+    assert which(executables).split(os.sep)[-1] in ['ls', 'mv']
+
+
+def test_py2ncl():
+    """Test for _py2ncl func."""
+    ncl_text = _py2ncl(None, 'tas')
+    assert ncl_text == 'tas = _Missing'
+    ncl_text = _py2ncl('cow', 'tas')
+    assert ncl_text == 'tas = "cow"'
+    ncl_text = _py2ncl([1, 2], 'tas')
+    assert ncl_text == 'tas = (/1, 2/)'
+    ncl_text = _py2ncl({'cow': 22}, 'tas')
+    assert ncl_text == 'tas = True\ntas@cow = 22\n'
+    with pytest.raises(ValueError) as ex_err:
+        _py2ncl([1, "cow"], 'tas')
+    assert 'NCL array cannot be mixed type:' in str(ex_err.value)
+    with pytest.raises(ValueError) as ex_err:
+        _py2ncl({"cow": 22}, None)
+    assert 'NCL does not support nested dicts:' in str(ex_err.value)
+
+
+def test_base_task():
+    """Test BaseTask basic attributes."""
+    tasks = set()
+    task = BaseTask(
+        name='task0',
+        ancestors=[BaseTask(name=f'task0-ancestor{j}') for j in range(2)],
+    )
+    for task0 in task.flatten():
+        task0.priority = 0
+    tasks.add(task)
+    task_names = [task.name for task in tasks]
+    assert task_names == ['task0']
+    ancestor_names = [anc.name for anc in list(tasks)[0].ancestors]
+    assert ancestor_names == ['task0-ancestor0', 'task0-ancestor1']
+    task = list(tasks)[0]
+    assert task.products == set()
+    assert task.output_files is None
+    assert task.activity is None
+    assert task.priority == 0
+
+
+def test_py_diagnostic_task_basics(tmp_path):
+    """Test DiagnosticTask basic attributes."""
+    diag_output_dir = tmp_path / 'mydiag'
+    diag_run_dir = diag_output_dir / 'run_dir'
+    diag_settings = {'run_dir': diag_run_dir, 'profile_diagnostic': False}
+
+    # diag exists
+    diag_script = tmp_path / 'diag_cow.py'
+    with open(diag_script, "w") as fil:
+        fil.write("import os\n\nprint(os.getcwd())")
+    tasks = set()
+    task = DiagnosticTask(
+        name='task0',
+        ancestors=[BaseTask(name=f'task0-ancestor{j}') for j in range(2)],
+        script=diag_script,
+        settings=diag_settings,
+        output_dir=diag_output_dir,
+    )
+    for task0 in task.flatten():
+        task0.priority = 0
+    tasks.add(task)
+    task_names = [task.name for task in tasks]
+    assert task_names == ['task0']
+    ancestor_names = [anc.name for anc in list(tasks)[0].ancestors]
+    assert ancestor_names == ['task0-ancestor0', 'task0-ancestor1']
+    task = list(tasks)[0]
+    assert task.script == diag_script
+    assert task.settings == diag_settings
+    assert task.output_dir == diag_output_dir
+
+    # diag doesn't exist
```

Comment:
Code below here should probably go into a separate test

Reply:
Was about to but then I got distracted with smth else :grin: